File: geral/livro_entendendo_algoritmos/selectionsort/selectionsort.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

func main() {
	values := load()

	values = selectionSort(values)

	fmt.Println(values)
}

func load() []int {
	file, err := os.Open("unsorted.txt")

	if err != nil {
		log.Fatal(err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)

	values := make([]int, 0)

	for scanner.Scan() {
		value, _ := strconv.Atoi(scanner.Text())

		values = append(values, value)
	}

	return values
}

func selectionSort(values []int) []int {
	sortedList := make([]int, 0)

	for len(values) != 0 {
		lowestValue, lowestIndex := lowest(values)

		sortedList = append(sortedList, lowestValue)

		values = append(values[:lowestIndex], values[lowestIndex+1:]...)
	}

	return sortedList
}

func lowest(values []int) (int, int) {
	lowestValue, lowestIndex := values[0], 0

	for index, value := range values {
		if lowestValue > value {
			lowestValue, lowestIndex = value, index
		}
	}

	return lowestValue, lowestIndex
}

File: geral/livro_entendendo_algoritmos/selection_sort.rb
def main()
    values = load

    values = selectionSort values

    puts values
end

def load()
    values = []

    IO.foreach("unsorted.txt") { |line| values.push line.to_i }

    values
end

def selectionSort(values)
    sorted_list = []

    while values.size != 0
        lowest_value, lowest_index = lowest values

        sorted_list.push lowest_value

        values.delete_at lowest_index
    end

    return sorted_list
end

def lowest(values)
    lowest_value, lowest_index = values[0], 0

    values.each_with_index do |value, index|
        if lowest_value > value
            lowest_value, lowest_index = value, index
        end
    end

    return lowest_value, lowest_index
end

main
